Refactor id functions

slgraham opened the pull request:
- simplify and shorten hat id functions
- refactor them into a new inherited contract
- use ide component constants for readability

nintynick commented:
variable names are confusing to me, and having them defined in the utility contract makes it a little harder to decipher too.

proposing changing names to something like this:
TOPHAT_BITS => TOP_LEVEL_ID_SPACE
LEVEL_BITS => ID_SPACE_PER_LEVEL
HAT_TREE_DEPTH = MAX_LEVELS

additionally I think a comment about addressable ID scheme that contextualizes these constant definitions would be helpful.
    
in terms of coverage, how's our test suite holding up? seems like it has been really useful so far, but want to make sure we're not introducing new edge cases we didn't yet design for. happy to dive in together.

slgraham commented:
For (1), adding more comments and natspec.

For (2), I originally used `HAT_TREE_DEPTH` to mirror the `MaxLevelsReached` error, but I think `MAX_LEVELS` is clearer so I'm going to change the error as well. I've also now factored out all functions that use these constants into HatsIdUtilities.sol, which should help with the deciphering.

For (3), agreed; changed to `_newHat`.

For (4), it's holding up alright. I don't believe we've added new edge cases with our recent changes. That said, the suite could probably be more comprehensive overall. Would be great to pair on identifying where to proliferate.

nintynick reviewed (approved):
natspec looks great! strong work. I think our future collaborators (including our future selves) will thank us for this. approving + merging.
